### PR TITLE
MD5 only when necessary, and following middleware chain

### PR DESCRIPTION
**Pull request by pat**
Um. @app.call is kinda important when it comes to middleware - guessing you're not using the alpha release at all? Or have I missed something?

### CODE REVIEW

**Comment by pat** [comment]
Added a third commit - handles query parameter values for allowed parameter keys when creating a Canonical URL. Useful when cancelling multi-part uploads.

Generally on the topic of multi-part uploads: I may have something working with this commit... first test was successful in bits and pieces, and currently have it running on a ~600MB file via a neater class. I'll confirm tomorrow and share the code.

**Comment by pat** [comment]
Here's some working code: https://gist.github.com/pat/7484995

File: lib/awsraw/s3/faraday_middleware.rb [hunk 1]
```diff
@@ -32,8 +32,9 @@ def call(env)
         )
 
         env[:request_headers]['Authorization'] = Signature.authorization_header(string_to_sign, @credentials)
-      end
 
+        @app.call(env)
```
Comment (lukearndt) [inline lib/awsraw/s3/faraday_middleware.rb:36]: Thanks @pat ! Looks like that got lost somewhere along the line.

Comment (notahat) [inline lib/awsraw/s3/faraday_middleware.rb:36]: Yep, was in there, and I nuked it by accident. We're using the older version that has it in.

Comment (pat) [inline lib/awsraw/s3/faraday_middleware.rb:36]: Ah :) Was wondering how the alpha had missed it!

**Pull request merged**